staking/programs/integrity-pool/src/state/pool.rs

```diff
@@ -102,54 +102,68 @@ impl PoolData {
         self.assert_up_to_date(current_epoch)?;
 
         let publisher_index = self.get_publisher_index(publisher)?;
-        let mut last_event_index: usize = self.num_events as usize;
+
         let mut delegator_reward: frac64 = 0;
         let mut publisher_reward: frac64 = 0;
-        loop {
-            // prevent infinite loop and double counting events
-            // by breaking the loop when visiting all events
-            if self.num_events as usize == last_event_index + MAX_EVENTS {
-                break;
-            }
 
-            match last_event_index {
-                0 => break,
-                _ => last_event_index -= 1,
-            }
+        let mut event_amounts = [0_u64; MAX_EVENTS];
 
-            let event = self.get_event(last_event_index);
-            if event.epoch < from_epoch {
-                break;
-            }
+        for i in 0..positions.get_position_capacity() {
+            let position = match positions.read_position(i)? {
+                Some(position) => position,
+                None => continue,
+            };
 
-            let mut amount = 0_u64;
-            for i in 0..positions.get_position_capacity() {
-                let position = positions.read_position(i)?;
-                if let Some(position) = position {
-                    let position_state =
-                        position.get_current_position(event.epoch, UNLOCKING_DURATION)?;
-                    if matches!(
-                        position.target_with_parameters,
-                        TargetWithParameters::IntegrityPool {
-                            publisher: ref position_publisher
-                        } if position_publisher == publisher
-                    ) && (position_state == PositionState::LOCKED
-                        || position_state == PositionState::PREUNLOCKING)
-                    {
-                        amount += position.amount;
+            let mut last_event_index: usize = self.num_events as usize;
+            loop {
+                // prevent infinite loop and double counting events
+                // by breaking the loop when visiting all events
+                if self.num_events as usize == last_event_index + MAX_EVENTS {
+                    break;
+                }
+
+                match last_event_index {
+                    0 => break,
+                    _ => last_event_index -= 1,
+                }
+
+                let event = self.get_event(last_event_index);
+                if event.epoch < from_epoch {
+                    break;
+                }
+
+                let position_state =
+                    position.get_current_position(event.epoch, UNLOCKING_DURATION)?;
+
+                match position.target_with_parameters {
```

Comment:
This doesn't need to happen in this for loop

Reply:
nice, this will reduce the CU even more.